code/datums/components/manual_heart.dm

```diff
@@ -0,0 +1,170 @@
+/datum/component/manual_heart/proc/on_pump(mob/owner)
+	last_pump = world.time
+	playsound(owner,'sound/effects/singlebeat.ogg', 40, TRUE)
+	owner.balloon_alert(owner, "your heart beats")
+
+	if(!iscarbon(owner))
+		return
+	var/mob/living/carbon/carbon_owner = owner
+
+	if(HAS_TRAIT(carbon_owner, TRAIT_NOBLOOD))
+		return
+	carbon_owner.blood_volume = min(carbon_owner.blood_volume + (blood_loss * 0.5), BLOOD_VOLUME_MAXIMUM)
+	carbon_owner.remove_client_colour(/datum/client_colour/manual_heart_blood)
+	add_colour = TRUE
+	carbon_owner.adjustBruteLoss(-heal_brute)
+	carbon_owner.adjustFireLoss(-heal_burn)
+	carbon_owner.adjustOxyLoss(-heal_oxy)
+
+/datum/component/manual_heart/process()
+	var/mob/living/carbon/carbon_parent = parent
+
+	//If they aren't connected, don't kill them.
+	if(!istype(carbon_parent) || !carbon_parent.client)
+		last_pump = world.time
+		return
+
+	if(world.time <= (last_pump + pump_delay))
+		return
+
+	carbon_parent.blood_volume = max(carbon_parent.blood_volume - blood_loss, 0)
+	to_chat(carbon_parent, span_userdanger("You have to keep pumping your blood!"))
+	last_pump = world.time - (2 SECONDS) //give two full seconds before losing more blood
+	if(add_colour)
+		carbon_parent.add_client_colour(/datum/client_colour/manual_heart_blood)
+		add_colour = FALSE
+
+///If a new heart is added, start processing.
+/datum/component/manual_heart/proc/check_added_organ(mob/organ_owner, obj/item/organ/new_organ)
+	SIGNAL_HANDLER
+
+	var/obj/item/organ/internal/heart/new_heart = new_organ
+
+	if(istype(new_heart) && check_valid())
+		last_pump = world.time
+		pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS)
+		START_PROCESSING(SSdcs, src)
+
+///If the heart is removed, stop processing.
+/datum/component/manual_heart/proc/check_removed_organ(mob/organ_owner, obj/item/organ/removed_organ)
+	SIGNAL_HANDLER
+
+	var/obj/item/organ/internal/heart/removed_heart = removed_organ
+
+	if(istype(removed_heart))
+		pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS)
+		STOP_PROCESSING(SSdcs, src)
+
+///Helper proc to check if processing can be restarted.
+/datum/component/manual_heart/proc/check_valid()
+	var/mob/living/carbon/carbon_parent = parent
+	if(!istype(parent))
+		return FALSE
+	var/obj/item/organ/internal/heart/parent_heart = carbon_parent.get_organ_slot(ORGAN_SLOT_HEART)
+	return (parent_heart && !HAS_TRAIT(carbon_parent, TRAIT_NOBLOOD) && carbon_parent.stat != DEAD)
+
+///Action to pump your heart. Cooldown will always be set to 1 second less than the pump delay.
+/datum/action/cooldown/manual_heart
+	name = "Pump your blood"
+	cooldown_time = 2 SECONDS
+	check_flags = NONE
+	button_icon = 'icons/obj/medical/organs/organs.dmi'
+	button_icon_state = "cursedheart-off"
+
+/datum/action/cooldown/manual_heart/Activate(atom/atom_target)
+	. = ..()
+
+	var/datum/component/manual_heart/heart = target
+	if(!istype(heart))
+		CRASH("Manual heart pump action created without corresponding component!")
+	heart.on_pump(owner)
```

Comment:
Don't know if I love this action calling a proc directly on a component.
I think we should either pass this in as a callback or send a HEART_MANUALLY_PULSED signal on the mob which triggers the `on_pump` proc.
Frankly you could probably even move most of the behaviour of that proc into this action, so long as the relevant parameters are passed down.